### DISCUSSION TITLE
New Function Code scheme proposal

### DISCUSSION THREAD

**Discussion by Miq1** [eModbus/eModbus#188] [Ideas]
Recently we had the request to allow a non-standard function code to be used, that was (illegally, in the sense of the Modbus standard) used by some server device in the wild.

While I found a solution to make that possible, I did not like what was required to do so in respect to cleanliness of the library code. So I started thinking about user-definable function codes and how to leave that completely in the user domain without affecting the library.

I came up with a new FunctionCode data class that is a bit "special", so I am asking for your opinions about it.

There will be the known standard function codes like READ_HOLD_REGISTER, WRITE_SINGLE_COIL etc. as before, optically not different from the existing. The difference is, the new codes are of class type FunctionCode, no more an ``enum`` of ``uint8_t`` values. 

This has some implications on using these, the type will be checked more strictly with the ``addRequest``, ``generateMessage`` etc. calls, no more simple number will be possible as a function code.
*That may break existing code that relies on the type identity of a function code and a ``uint8_t``.*

On the other hand new, user-domain function codes may be defined in an application, that will be handled by the library exactly as if these would have been standard ones. A new code is defined like
```
FunctionCode& MyOwnSpecialCode NewFC<112>::instance();
```
The name ``MyOwnSpecialCode`` now is a reference to the code with numer 112. This will be unique and fixed from now on, any attempt to redefine the same code will end in another named reference to the very same code. Hence
```
FunctionCode& NotDifferent NewFX<112>::instance();
```
anywhere in the code will just return the same reference that ``MyOwnSpecialCode`` holds, it will be two names for the same entity.

What do you think? Is that worthwhile integrating into the lib?

**Comment by bertmelis** [comment]
As for backwards compatibility, there is only a change when you tried to have custom function codes. In 99% of the cases, user code will not change, right? --> I have no objections here

Whether it is worthwhile or not, who shall say? You already give it a thorough thought so the majority of the work is already done 😄 .

**Reply by Miq1** [reply]
As I wrote: most code, if using the symbolic names for FCs, will continue to work after a recompile. Those applications using plain numbers for FCs will not.

**Reply by bertmelis** [reply]
That's a change I'm very willing to allow. Never been a fan of plain numbers without meaning in code.

**Reply by Miq1** [reply]
Okay, my opinion as well. Thanks.

**Comment by Miq1** [comment]
Declined due to unforeseen side effects.
